perf/evp_fetch.c

```diff
@@ -0,0 +1,212 @@
+            if (kdf == NULL) {
+                fprintf(stderr, "Failed to fetch %s\n", fetch_entries[j].alg);
+                err = 1;
+                return;
+            }
+            EVP_KDF_free(kdf);
+            break;
+        case FETCH_MAC:
+            EVP_MAC *mac = EVP_MAC_fetch(ctx, fetch_entries[j].alg,
+                                         fetch_entries[j].propq);
+            if (mac == NULL) {
+                fprintf(stderr, "Failed to fetch %s\n", fetch_entries[j].alg);
+                err = 1;
+                return;
+            }
+            EVP_MAC_free(mac);
+            break;
+        case FETCH_RAND:
+            EVP_RAND *rnd = EVP_RAND_fetch(ctx, fetch_entries[j].alg,
+                                           fetch_entries[j].propq);
+            if (rnd == NULL) {
+                fprintf(stderr, "Failed to fetch %s\n", fetch_entries[j].alg);
+                err = 1;
+                return;
+            }
+            EVP_RAND_free(rnd);
+            break;
+        default:
+            err = 1;
+            return;
+        }
+    }
+    end = ossl_time_now();
+    times[num] = ossl_time_subtract(end, start);
+}
+
+int main(int argc, char *argv[])
+{
+    OSSL_TIME duration;
+    OSSL_TIME ttime;
+    double av;
+    int terse = 0;
+    int argnext;
+    size_t i;
+    int rc = EXIT_FAILURE;
+
+    if ((argc != 2 && argc != 3)
+                || (argc == 3 && strcmp("--terse", argv[1]) != 0)) {
+        printf("Usage: %s [--terse] threadcount\n", argv[0]);
+        return EXIT_FAILURE;
+    }
+
+    if (argc == 3) {
+        terse = 1;
+        argnext = 2;
+    } else {
+        argnext = 1;
+    }
+
+    threadcount = atoi(argv[argnext]);
+    if (threadcount < 1) {
+        printf("threadcount must be > 0\n");
+        return EXIT_FAILURE;
+    }
+
+    ctx = OSSL_LIB_CTX_new();
+    if (ctx == NULL)
+        return EXIT_FAILURE;
+
+    times = OPENSSL_malloc(sizeof(OSSL_TIME) * threadcount);
+    if (times == NULL) {
+        printf("Failed to create times array\n");
+        return EXIT_FAILURE;
+    }
+
+    if (!perflib_run_multi_thread_test(do_fetch, threadcount, &duration)) {
+        printf("Failed to run the test\n");
+        goto out;
+    }
+
+    if (err) {
+        printf("Error during test\n");
+        goto out;
+    }
+
+    ttime = times[0];
+    for (i = 1; i < threadcount; i++)
+        ttime = ossl_time_add(ttime, times[i]);
+
+    /*
+     * EVP_PKEY_new_raw_public_key is pretty fast, running in
+     * only a few us.  But ossl_time2us does integer division
+     * and so because the average us computed above is less than
+     * the value of OSSL_TIME_US, we wind up with truncation to
+     * zero in the math.  Instead, manually do the division, casting
+     * our values as doubles so that we compute the proper time
+     */
+    av = ((double)ossl_time2ticks(ttime) / (double)NUM_CALLS_PER_TEST) /(double)OSSL_TIME_US;
```

Comment:
Stylistic: only the first needs to be there, the rest will switch to double automatically.

Reply:
I can make this change, but I would want to do it in all the tests, as they follow this pattern.  suggest a separate pull request to do that to avoid feature creep here.

Reply:
No problem.  It was only a stylistic suggestion.